src/Microsoft.Data/PrimitiveColumnContainer.cs

```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
+                lastNullBitMapBuffer.Length = allocatable;
                 length -= allocatable;
                 Length += lastBuffer.Length;
             }
         }
 
-        public void Append(T value)
+        public void Append(T? value)
         {
             if (Buffers.Count == 0)
             {
                 Buffers.Add(new DataFrameBuffer<T>());
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
             DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
             if (lastBuffer.Length == lastBuffer.MaxCapacity)
             {
                 lastBuffer = new DataFrameBuffer<T>();
+                Buffers.Add(lastBuffer);
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
-            lastBuffer.Append(value);
+            lastBuffer.Append(value ?? default);
+            SetBit(Length, value.HasValue ? true : false);
             Length++;
         }
 
+        public bool IsValid(long index) => NullCount == 0 || GetBit(index);
+
+        private void SetBit(long index, bool value)
+        {
+            if (index < 0 || index > Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            // First find the right bitMapBuffer
+            int bitMapIndex = (int)(index / Buffers[0].MaxCapacity);
+            Debug.Assert(NullBitMapBuffers.Count > bitMapIndex);
+            DataFrameBuffer<byte> bitMapBuffer = NullBitMapBuffers[bitMapIndex];
+
+            // Set the bit
+            index -= bitMapIndex * Buffers[0].MaxCapacity;
+            int bitMapBufferIndex = (int)index / 8;
+            Debug.Assert(bitMapBuffer.Length >= bitMapBufferIndex);
+            if (bitMapBuffer.Length == bitMapBufferIndex)
+                bitMapBuffer.Append(0);
+            byte curBitMap = bitMapBuffer[bitMapBufferIndex];
+            byte newBitMap;
+            if (value)
+            {
+                newBitMap = (byte)(curBitMap | (byte)(1 << (int)(index % 8)));
+                if ((curBitMap >> ((int)(index % 8)) & 1) == 0 && index < Length && NullCount > 0)
+                {
+                    // Old value was null.
+                    NullCount--;
+                }
+            }
+            else
+            {
+                if ((curBitMap >> ((int)(index % 8)) & 1) == 1 && index < Length)
+                { 
+                    // old value was NOT null and new value is null
+                    NullCount++;
+                }
+                else if (index == Length)
+                {
+                    // New entry from an append
+                    NullCount++;
+                }
+                newBitMap = (byte)(curBitMap & (byte)~(1 << (int)(index % 8)));
+            }
+            bitMapBuffer[bitMapBufferIndex] = newBitMap;
+        }
+
+        private bool GetBit(long index)
+        {
+            if (index < 0 || index > Length)
```

Comment:
```suggestion
            if ((ulong)index > Length)
```
See above. Also for ThrowHelper-thoughts.